docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+
+- Create a new field Spec.ControlPlane
+    - Move Spec.ControlPlaneEndpoint to Spec.ControlPlane.Endpoint
+    - Move Spec.ControlPlaneRef to Spec.ControlPlane.ProviderRef
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha4
+kind: Cluster
+metadata:
+  name: my-cluster
+spec:
+  controlPlane:
+    endpoint:
+      host: test.example.com
+      port: 443
+    providerRef:
+      apiGroup: cluster.x-k8s.io
+      kind: KubeadmControlPlane
+      name: my-cluster-control-plane
+```
+
+#### Changes to Cluster reconciliation:
+
+![Cluster Reconciliation Flow](images/load-balancer/cluster-reconciliation.png)
+
+##### Scenario: externally managed endpoint and control plane
+
+User wants to provide an externally managed endpoint for use when creating a cluster to a pre-existing control plane that is not managed by Cluster API.
+
+Preconditions:
+- `Spec.ControlPlane.Endpoint` is non-empty
+- `Spec.ControlPlane.ProviderRef` is nil
+- A pre-configured kubeconfig secret that can access the pre-existing control plane exists with the expected name.
+
+Reconciliation:
+- Set `ControlPlaneInitializedCondition` to True
+- Test availability of the controlplane using the configured endpoint and provided kubeconfig:
+  - If available: Set `Status.ControlPlaneReady` and the `ControlPlaneReadyCondition` to True
+  - Otherwise: Set `Status.ControlPlaneReady` and the `ControlPlaneReadyCondition` to False
+
+##### Scenario: externally managed endpoint
+
+User wants to provide an externally managed endpoint for use when creating a cluster, but wants Cluster API to manage the control plane for the cluster.
+
+Preconditions:
+- `Spec.ControlPlane.Endpoint` is non-empty
+- `Spec.ControlPlane.ProviderRef` is defined
+
+Reconciliation:
+- Fetch the referenced provider CR as `cp`
+  - If `cp.Spec.Endpoint` is empty, set `cp.Spec.Endpoint` to `Spec.ControlPlane.Endpoint`
+  - If `cp.Spec.Endpoint` is non-empty and is not equal to `Spec.ControlPlane.Endpoint`, error
+  - If `cp.Spec.Endpoint` is non-empty and is equal to `Spec.ControlPlane.Endpoint`:
+    - Set `Status.ControlPlaneReady` and conditions based on state of `cp`
+
+##### Scenario: Cluster API managed endpoint
+
+User wants Cluster API to manage the endpoint for the cluster control plane.
+
+Preconditions:
+- `Spec.ControlPlane.Endpoint` is empty
+- `Spec.ControlPlane.ProviderRef` is defined
+
+Reconciliation:
+- Fetch the referenced provider CR as `cp`
+  - If `cp.Spec.Endpoint` is empty, return early and wait for triggered re-reconciliation
+  - If `cp.Spec.Endpoint` is non-empty and is not equal to `Spec.ControlPlane.Endpoint`, error
+  - If `cp.Spec.Endpoint` is non-empty and is equal to `Spec.ControlPlane.Endpoint`:
+    - Set `Status.ControlPlaneReady` and conditions based on state of `cp`
+
+#### Changes to ControlPlane provider types:
+
+- Add spec.Endpoint, if not defined it will be populated by the load balancer provider (if defined)
+- Add optional LoadBalancerRef (if needed, based on the provider)
```

Comment:
This is the part where I still need to properly understand this proposal.
In my mind, the load balancer should sit at Cluster level, fill an endpoint, and the control plane should use it.
Instead here we are modelling the load balancer as a detail of the control plane. It will be great id we can provide a little bit more context of why we are choosing this design.

Reply:
Thanks for the detailed explanation.
I think that documenting that we are modelling the load balancer as a detail fo the control plane in order to provide better support form managed control plane solutions (like EKS) will be great